R/source.R

```diff
@@ -12,6 +12,12 @@
 #'    Literal data is most useful for examples and tests. It must contain at
 #'    least one new line to be recognised as data (instead of a path).
 #' @param skip Number of lines to skip before reading data.
+#' @param comment A string or character vector used to identify comments. Any
+#'                text after the comment characters will be silently ignored.
+#' @param encoding The text encoding of the data given in `file`. If `NULL` is
```

Comment:
At first I thought that having "UTF-8" as default was the best approach. This is what `default_locale()$encoding` gives, so it makes sense to keep it everywhere as default.

However, when working with datasources that are literal strings (like `"à b\n1 2\n3 4"`) I found it made more sense to (by default) autodetect the encoding using `Encoding()` if possible. This forced me to change the default value from `"UTF-8"` to `NULL` and autodetect it when a literal string is passed.

Then there was `datasource_connection`. If for instance the connection is a pipe, the default encoding usually is the one given by the locale. Here I assumed "bytes", meaning "something ASCII compatible that I will not convert". If I had chosen "UTF-8" then errors like "invalid UTF-8 character" would appear when someone used a pipe giving non-UTF8 characters, breaking backwards compatibility in those cases.

For `datasource_raw` I used a similar reasoning: "bytes" does not check for UTF-8 validity and most encodings are ASCII compatible, so it will preserve better backwards compatibility.

Having a default value for each type of datasource maybe complicates things too much, as it then forced me to check other functions calling `datasource()` and set the new default `NULL` encoding if it was not given.

If you prefer I can simplify this, keeping "UTF-8" as the only default in all cases.